inc/c_util/doublylinkedlist.h

```diff
@@ -4,26 +4,61 @@
 #ifndef DOUBLYLINKEDLIST_H
 #define DOUBLYLINKEDLIST_H
 
-#include "umock_c/umock_c_prod.h"
-
 #ifdef __cplusplus
-extern "C"
-{
+#else
+#include <stdbool.h>
 #endif
 
+#include "macro_utils/macro_utils.h"
+
 typedef struct DLIST_ENTRY_TAG
 {
     struct DLIST_ENTRY_TAG *Flink;
     struct DLIST_ENTRY_TAG *Blink;
 } DLIST_ENTRY, *PDLIST_ENTRY;
 
+#define DLIST_FIND_RESULT_VALUES \
+    DLIST_FIND_RESULT_FOUND, \
+    DLIST_FIND_RESULT_NOT_FOUND, \
+    DLIST_FIND_RESULT_ERROR
+
+MU_DEFINE_ENUM(DLIST_FIND_RESULT, DLIST_FIND_RESULT_VALUES);
+
+#define DLIST_MATCH_FUNCTION_RESULT_VALUES \
+    DLIST_MATCH_FUNCTION_MATCHING, \
+    DLIST_MATCH_FUNCTION_NOT_MATCHING, \
+    DLIST_MATCH_FUNCTION_ERROR
+
+MU_DEFINE_ENUM(DLIST_MATCH_FUNCTION_RESULT, DLIST_MATCH_FUNCTION_RESULT_VALUES);
+
+#define DLIST_CONDITION_FUNCTION_RESULT_VALUES \
+    DLIST_CONDITION_FUNCTION_TRUE, \
+    DLIST_CONDITION_FUNCTION_FALSE, \
+    DLIST_CONDITION_FUNCTION_ERROR
+
+MU_DEFINE_ENUM(DLIST_CONDITION_FUNCTION_RESULT, DLIST_CONDITION_FUNCTION_RESULT_VALUES);
+
+typedef DLIST_MATCH_FUNCTION_RESULT (*DLIST_MATCH_FUNCTION)(PDLIST_ENTRY listEntry, const void* matchContext);
+typedef DLIST_CONDITION_FUNCTION_RESULT (*DLIST_CONDITION_FUNCTION)(PDLIST_ENTRY listEntry, const void* conditionContext, bool* continueProcessing);
```

Comment:
>DLIST_CONDITION_FUNCTION [](http://example.com/codeflow?start=42&length=24)

`DLIST_CONDITION_FUNCTION` returns DLIST_CONDITION_FUNCTION_TRUE/FALSE/ERROR.

Can `DLIST_CONDITION_FUNCTION` return TRUE and have \*continueProcessing set to "false"?

Can `DLIST_CONDITION_FUNCTION` return `DLIST_CONDITION_FUNCTION_FALSE` and *continueProcessing set to "true"?

Reply:
1) Yes. The current node will be removed and DList_RemoveIf will stop iterating over the list.

2) Yes. The current node will not be removed and DList_RemoveIf will continue iterating over the list.